core-services/src/main/java/org/zalando/nakadi/validation/ResourceValidationHelperService.java

```diff
@@ -0,0 +1,148 @@
+
+    static final int MAX_ANNOTATION_LENGTH = 1000;
+    static final int MAX_LABEL_LENGTH = 63;
+
+    private static final String ERROR_MESSAGE_TEMPLATE = "Error validating <%s:%s>; %s";
+    private static final String EMPTY_KEY_ERROR = "Key cannot be empty.";
+    private static final String MULTI_PREFIX_ERROR = "Key cannot have multiple prefixes.";
+    private static final String EMPTY_KEY_PREFIX_ERROR = "Key prefix cannot be empty.";
+    private static final String EMPTY_KEY_NAME_ERROR = "Key name cannot be empty.";
+    private static final String LONG_KEY_NAME_ERROR = "Key name cannot be more than 63 characters.";
+    private static final String START_END_KEY_NAME_ERROR = "Key name should start and end with [a-zA-Z0-9].";
+    private static final String START_END_LABEL_ERROR = "Label should start and end with [a-zA-Z0-9].";
+    private static final String KEY_NAME_CHARACTER_ERROR = "Key name should use only [a-zA-Z0-9] and `-`, '_', and `.`";
+    private static final String LABEL_CHARACTER_ERROR = "Label should use only [a-zA-Z0-9] and `-`, '_', and `.`";
+    private static final String LONGER_PREFIX_ERROR = "Key prefix cannot be more than 253 characters.";
+    private static final String INVALID_PREFIX_ERROR = "Key prefix must follow RFC 1123 DNS subdomain format. \n" +
+            "It must contain one or more segments separated by '.'(dot), and each segment must start and end with " +
+            "[a-z0-9], and may contain - (hyphen) and [a-z0-9] in between.";
+    private static final String PREFIX_SEPARATOR = "/";
+    private static final String LONG_ANNOTATION_ERROR = "Annotation cannot be more than " + MAX_ANNOTATION_LENGTH
+            + " characters long.";
+    private static final String LONG_LABEL_ERROR = "LABEL cannot be more than " + MAX_LABEL_LENGTH + "charecters long.";
+
+    public void checkAnnotations(final ResourceAnnotations annotations) throws InvalidResourceAnnotationException {
+        if (annotations != null) {
+            annotations.forEach(this::validateAnnotation);
+        }
+    }
+
+    public void checkLabels(final ResourceLabels labels) throws InvalidResourceLabelException {
+        if (labels != null) {
+            labels.forEach(this::validateLabels);
+        }
+    }
+
+    private void validateAnnotation(final String key, final String value) {
+        validateKey(key).ifPresent(error -> {
+            throw new InvalidResourceAnnotationException(String.format(ERROR_MESSAGE_TEMPLATE, key, value, error));
+        });
+        if (value != null && value.length() > MAX_ANNOTATION_LENGTH) {
+            throw new InvalidResourceAnnotationException(
+                    String.format(ERROR_MESSAGE_TEMPLATE, key, value, LONG_ANNOTATION_ERROR));
+        }
+    }
+
+    private void validateLabels(final String key, final String value) {
+        validateKey(key).ifPresent(error -> {
+            throw new InvalidResourceLabelException(error);
```

Comment:
I think we should also use `ERROR_MESSAGE_TEMPLATE` here.  But maybe the template should be extended to also report resource type, e.g. "error validating %s", where we put either "annotation" or "label".

Reply:
Did this in a uniform way with template, but using the exception class.